test-suite/output/bug_16262.v

```
@@ -0,0 +1,10 @@
+Declare Scope foo.
+Check (nat -> nat).
+Notation "a -> b" := (a,b)
+   (b at level 200, only printing, right associativity, at level 99,
+   format "a -> b") : foo.
```

This used to issue a warning
```
Warning: Notation "_ -> _" was already defined with a different format.
[notation-incompatible-format,parsing,default]
```
I'd say we should preserve the warning rather than silently ignoring the format.

---

Is it ignored? The test should probably check how `(1, 2)` prints (should be something like `1->2` after this command IIUC)

---

No it's not ignored.
~~~coq
Declare Scope foo.
Open Scope foo.

Notation "a -> b" := (a,b)
   (b at level 200, only printing, right associativity, at level 99,
   format "a -> b") : foo.

Check (1, 2).
(* 1->2 *)
~~~
and this can't be done in Reserved Notation because it's specific to this interpretation.
Does Reserved Notation with only printing even make sense?

---

> No it's not ignored.

Before the PR indeed. IIUC that's exactly what the current PR is changing.

> and this can't be done in Reserved Notation because it's specific to this interpretation.

Well, that's what's done two line belows. That's the whole point for `Reserved Notation`: setting parsing/printing params (levels and format) that are tied to the notation itself, not to any of its interpretation (if you `Check Nat.add` after your example you will get `nat->nat->nat` (without spaces)).

> Does Reserved Notation with only printing even make sense?

Yes

---

>Before the PR indeed. IIUC that's exactly what the current PR is changing.

No, with the PR, after the Notation `nat -> nat` prints `nat -> nat` and `(0,1)` prints `0->1`. Before the PR `nat -> nat` also prints `nat->nat`.

---

That's not my understanding of the test output [test-suite/output/bug_16262.out](https://github.com/coq/coq/pull/16329/files#diff-c9148a41b458d273bb9a3aba878e48364278ff3bd3bdbbb53bea7c4d6cbf2c1d) @herbelin could you clarify please?

---

I don't know what your understanding is but I tested locally.

---

I just read the test and its output, maybe it's outdated then.

---

The test is up to date but incomplete (doesn't test printing `(1,2)`).

---

Ha, got it thanks. The test should indeed be completed.
So IIUC this PR introduces a notion of format attached to interpretations of printing only notations. That's a nice feature but I'm not so sure we want it. It doesn't go in the direction of making the parsing/printing vs interpretation separation easier to understand. Is there a reasonable use case for that feature?

---

I added a test of `(1,2)`.

The motivation originally comes from #16262. It is convenient to have a generic format that applies to all printing of a notation string, but nothing forces per se an interpretation for this notation string to share the same format as another interpretation for the same notation string. So, to change the generic format, only `Reserved Notation` or parsing-enabled `Notation` can be used.

I don't know in which sense you say that this does not go in the direction of making the parsing/printing vs interpretation separation easier to understand.

For instance, we could ask for more, that is that only `Reserved Notation` can change the generic format, and that, otherwise a new parsing-printing-enabled `Notation` changes only the format for this particular interpretation. I would actually do this change too (but depending first on your answer).

> More generally, I tend to think that enabling tinkering with format and levels in `Notation` rather than `Reserved Notation` was a bad design choice, as it doesn't help users distinguish the parsing/printing phase from the notation interpretation phase. But we won't fix that now.

I somehow agree :(, but the current situation (assuming the PR + the change proposed in the previous paragraph) can also be explained by saying that an initial `Notation string := term (syntax_modifiers, format)` is a short-hand for `Reserved Notation string (syntax_modifiers, format)`, where the format is generic, followed by `Notation string := term`, and a second `Notation string := term (syntax_modifiers, format)` is a short-hand for `Reserved Notation string (syntax_modifiers)`, overriding the modifiers, followed by `Notation string := term (format)`, where the format is specific.

---

> I added a test of `(1,2)`.

Thanks

> The motivation originally comes from #16262.

Sorry but I wouldn't call that a compelling example. I'd rather agree with @Blaisorblade there that it just looks like a misuse of the Notation command.

> It is convenient to have a generic format that applies to all printing of a notation string, but nothing forces per se an interpretation for this notation string to share the same format as another interpretation for the same notation string. So, to change the generic format, only `Reserved Notation` or parsing-enabled `Notation` can be used.

Sure, I do agree it's perfectly possible (and apparently pretty easy) to have per-scope formats for printing only notations. My question is more: do we want to provide that feature? Is it of any practical use? Wouldn't it be confusing to have the same notation print a bit differently in different contexts? If the feature is really needed, wouldn't it be better achieved with custom entries? I'm not saying the answer is no, just that it's not obvious to me.

> I don't know in which sense you say that this does not go in the direction of making the parsing/printing vs interpretation separation easier to understand.
> 
> For instance, we could ask for more, that is that only `Reserved Notation` can change the generic format, and that, otherwise a new parsing-printing-enabled `Notation` changes only the format for this particular interpretation.

This may help, but I think we should first answer the previous question anyway.
Maybe a way to better convey the locality of the format would be a specific syntax for them like `Notation "..." := (...) (only printing, local format "...") : bar_scope` (or `scope format`).

> I would actually do this change too (but depending first on your answer).

Probably material for a separate PR anyway (let's avoid PR blow-up).

> > More generally, I tend to think that enabling tinkering with format and levels in `Notation` rather than `Reserved Notation` was a bad design choice, as it doesn't help users distinguish the parsing/printing phase from the notation interpretation phase. But we won't fix that now.
> 
> I somehow agree :(, but the current situation (assuming the PR + the change proposed in the previous paragraph) can also be explained by saying that an initial `Notation string := term (syntax_modifiers, format)` is a short-hand for `Reserved Notation string (syntax_modifiers, format)`, where the format is generic, followed by `Notation string := term`, and a second `Notation string := term (syntax_modifiers, format)` is a short-hand for `Reserved Notation string (syntax_modifiers)`, overriding the modifiers, followed by `Notation string := term (format)`, where the format is specific.

I do agree, and I dream of enforcing that, by enabling to mention levels and format only in `Reserved Notation`, whereas scopes could be mentionned only in `Notation`. This would force users to first use `Reserve Notation` before `Notation`. I think that would help newcomers to realize that those are two distinct, successive, phases. Would also help realizing that, at least for parsing, levels are inherently tied to a reserved notation, rather than any of its interpretation.
My main reservation with the introduction of local format, as done in the current PR, is that this would break the above separation, since format would then also have to appear in `Notation`, rather than only in `Reserved Notation`. That's why I'd like to introduce the feature only if it's actually useful, and then I'd better find a way to limit the confusion between "local" and "global" formats.

---

Ok, if the feature already exist, maybe it's too late to ask such questions (although #10832 and linked issues there rather gave me the feeling that those local formats could be more confusing than requested by users).

> Forcing to use expressions like local format or scope format to clarify what we are doing is a very good idea.

Then let's do that (I have a small preference for `scope format` FWIW).

---

A possible drawback of `scope format` is that a specific format can be assigned also to a "lonely" notation but I don't have real alternatives though. Internally, it is called `specific format`, but that's a bit heavy. We could think of `proper format`, or `ad hoc format`?

In any case, here is a proposal:
- we introduce the new syntax, say `scope format`
- we emit a warning any time `format` is used in a situation where `scope format` is expected, telling to use `scope format` to deactivate the warning

Does this sound ok to you?

---

> A possible drawback of `scope format` is that a specific format can be assigned also to a "lonely" notation but I don't have real alternatives though. Internally, it is called `specific format`, but that's a bit heavy. We could think of `proper format`, or `ad hoc format`?

Maybe `scope-specific format`? (that's a tad lengthy though)

> In any case, here is a proposal:
> 
>     * we introduce the new syntax, say `scope format`
> 
>     * we emit a warning any time `format` is used in a situation where `scope format` is expected, telling to use `scope format` to deactivate the warning
> 
> Does this sound ok to you?

Sounds good (the warning can be a deprecation, mentioning Coq 8.20)

---

Is it tied to the scope? `specific format` sounds fine to me.
The doc says
>When a format is attached to a reserved notation (with the format syntax_modifier), it is used by default by all subsequent interpretations of the corresponding notation. Individual interpretations can override the format.

no mention of scopes

---

Right, it is only tied to the interpretation, so `specific format` sounds indeed better.

---

IIUC this side discussion isn't blocking for this PR ie it's ready to merge.